File: libraries/typescript/src/document.ts
import type {
  IfcxDocument as IIfcxDocument,
  Header, Tables, Entity, DrawingObject, BlockDefinition,
  Layer, TextStyle, DimStyle, Linetype,
} from './types.js';

/**
 * In-memory representation of an IFCX document.
 * Provides methods for creating, querying, and manipulating drawing data.
 */
export class IfcxDocument {
  ifcx: '1.0' = '1.0';
  header: Header;
  tables: Tables;
  blocks: Record<string, BlockDefinition>;
  entities: Entity[];
  objects: DrawingObject[];
  extensions: Record<string, unknown>;

  private nextHandle = 1;

  constructor(init?: Partial<IIfcxDocument>) {
    this.header = init?.header ?? { units: { measurement: 'metric', linear: 'millimeters' } };
    this.tables = init?.tables ?? { layers: { '0': {} }, linetypes: {}, textStyles: {}, dimStyles: {} };
    this.blocks = init?.blocks ?? {};
    this.entities = init?.entities ?? [];
    this.objects = init?.objects ?? [];
    this.extensions = init?.extensions ?? {};
  }

  /** Generate a unique hex handle */
  allocHandle(): string {
    return (this.nextHandle++).toString(16).toUpperCase();
  }

  /** Add a layer */
  addLayer(name: string, props: Layer = {}): void {
    this.tables.layers ??= {};
    this.tables.layers[name] = props;
  }

  /** Add a linetype */
  addLinetype(name: string, props: Linetype): void {
    this.tables.linetypes ??= {};
    this.tables.linetypes[name] = props;
  }

  /** Add a text style */
  addTextStyle(name: string, props: TextStyle): void {
    this.tables.textStyles ??= {};
    this.tables.textStyles[name] = props;
  }

  /** Add a dimension style */
  addDimStyle(name: string, props: DimStyle): void {
    this.tables.dimStyles ??= {};
    this.tables.dimStyles[name] = props;
  }

  /** Add an entity and auto-assign handle */
  addEntity(entity: Entity): string {
    const handle = this.allocHandle();
    entity.handle = handle;
    this.entities.push(entity);
    return handle;
  }

  /** Add a block definition */
  addBlock(block: BlockDefinition): void {
    this.blocks[block.name] = block;
  }

  /** Find entities by type */
  findByType<T extends Entity['type']>(type: T): Extract<Entity, { type: T }>[] {
    return this.entities.filter(e => e.type === type) as Extract<Entity, { type: T }>[];
  }

  /** Find entities on a specific layer */
  findByLayer(layer: string): Entity[] {
    return this.entities.filter(e => e.layer === layer);
  }

  /** Get entity by handle */
  getByHandle(handle: string): Entity | undefined {
    return this.entities.find(e => e.handle === handle);
  }

  /** Serialize to IFCX JSON */
  toJSON(): IIfcxDocument {
    return {
      ifcx: this.ifcx,
      header: this.header,
      tables: this.tables,
      blocks: this.blocks,
      entities: this.entities,
      objects: this.objects,
      extensions: this.extensions,
    };
  }

  /** Parse from IFCX JSON */
  static fromJSON(json: IIfcxDocument): IfcxDocument {
    return new IfcxDocument(json);
  }
}

File: libraries/rust/src/document.rs
//! In-memory IFCX document.

use crate::types::*;
use std::collections::HashMap;

/// In-memory representation of an IFCX document.
pub struct IfcxDocument {
    pub file: IfcxFile,
    next_handle: u64,
}

impl IfcxDocument {
    /// Create a new empty document.
    pub fn new() -> Self {
        Self {
            file: IfcxFile {
                ifcx: "1.0".to_string(),
                header: Header {
                    version: None,
                    author: None,
                    organization: None,
                    application: None,
                    units: Some(Units {
                        linear: Some("millimeters".to_string()),
                        measurement: Some("metric".to_string()),
                        extra: HashMap::new(),
                    }),
                    extents: None,
                    limits: None,
                    current_layer: Some("0".to_string()),
                    linetype_scale: None,
                    extra: HashMap::new(),
                },
                tables: Some(Tables {
                    layers: Some(HashMap::from([("0".to_string(), Layer {
                        color: None, linetype: None, lineweight: None,
                        frozen: None, locked: None, off: None, plot: None,
                    })])),
                    linetypes: Some(HashMap::new()),
                    text_styles: Some(HashMap::new()),
                    dim_styles: Some(HashMap::new()),
                    extra: HashMap::new(),
                }),
                blocks: Some(HashMap::new()),
                entities: Vec::new(),
                objects: None,
                extensions: None,
            },
            next_handle: 1,
        }
    }

    /// Generate a unique hex handle.
    pub fn alloc_handle(&mut self) -> String {
        let handle = format!("{:X}", self.next_handle);
        self.next_handle += 1;
        handle
    }

    /// Add an entity and auto-assign handle.
    pub fn add_entity(&mut self, mut entity: Entity) -> String {
        let handle = self.alloc_handle();
        entity.handle = Some(handle.clone());
        self.file.entities.push(entity);
        handle
    }

    /// Find entities by type.
    pub fn find_by_type(&self, entity_type: &str) -> Vec<&Entity> {
        self.file.entities.iter()
            .filter(|e| e.entity_type == entity_type)
            .collect()
    }

    /// Find entities on a layer.
    pub fn find_by_layer(&self, layer: &str) -> Vec<&Entity> {
        self.file.entities.iter()
            .filter(|e| e.layer.as_deref() == Some(layer))
            .collect()
    }
}

impl Default for IfcxDocument {
    fn default() -> Self {
        Self::new()
    }
}

impl From<IfcxFile> for IfcxDocument {
    fn from(file: IfcxFile) -> Self {
        Self { file, next_handle: 1 }
    }
}
